### PR TITLE
feat: add ShaderSystem#maxTextures

### PR DESCRIPTION
**Pull request by dev7355608**
##### Description of change

Added `ShaderSystem#maxTextures`, which is the maximum number of textures that can be used in a shader.

##### Pre-Merge Checklist

- [ ] Tests and/or benchmarks are included
- [ ] Documentation is changed or added
- [ ] Lint process passed (`npm run lint`)
- [ ] Tests passed (`npm run test`)

### CODE REVIEW

File: src/rendering/batcher/gl/utils/checkMaxIfStatementsInShader.ts [hunk 1]
```diff
@@ -38,23 +38,30 @@ export function checkMaxIfStatementsInShader(maxIfs: number, gl: GlRenderingCont
...
+            else
+            {
+                // valid!
+                break;
+            }
         }
     }
+    finally
+    {
+        gl.deleteShader(shader);
```
Comment (dev7355608) [inline src/rendering/batcher/gl/utils/checkMaxIfStatementsInShader.ts:63]: Just added the `deleteShader` here.

File: src/rendering/batcher/gl/utils/maxRecommendedTextures.ts [hunk 2]
```diff
@@ -23,5 +23,7 @@ export function getMaxTexturesPerBatch(): number
     maxTexturesPerBatchCache = checkMaxIfStatementsInShader(
         maxTexturesPerBatchCache, gl);
 
+    gl.getExtension('WEBGL_lose_context')?.loseContext();
```
Comment (dev7355608) [inline src/rendering/batcher/gl/utils/maxRecommendedTextures.ts:26]: Perhaps helps cleaning up the webgl context faster.

File: src/rendering/renderers/gpu/shader/GpuShaderSystem.ts [hunk 3]
```diff
@@ -24,13 +25,17 @@ export class GpuShaderSystem implements System
...
+
     private _gpu: GPU;
 
     private readonly _gpuProgramData: Record<number, GPUProgramData> = Object.create(null);
 
     protected contextChange(gpu: GPU): void
     {
         this._gpu = gpu;
+
+        this.maxTextures = getMaxTexturesPerBatch();
```
Comment (dev7355608) [inline src/rendering/renderers/gpu/shader/GpuShaderSystem.ts:38]: I suppose we could use webgpu here to determine the maximum textures given the webgpu context. But I don't know how that would work.

Comment (GoodBoyDigital) [reply src/rendering/renderers/gpu/shader/GpuShaderSystem.ts:38]: a good point, this turned out to be a pain as often max textures is required before a renderer is created. Figured WebGL would be a safest to just use in all cases!

**Review by GoodBoyDigital — APPROVED** [review]
nice!

File: src/rendering/renderers/gpu/shader/GpuShaderSystem.ts (hunk 3 above)
Comment (GoodBoyDigital) [reply src/rendering/renderers/gpu/shader/GpuShaderSystem.ts:38]: although here, we could do:
```gpu.device.limits.maxSampledTexturesPerShaderStage```

Comment (dev7355608) [reply src/rendering/renderers/gpu/shader/GpuShaderSystem.ts]: Made the change to use `gpu.device.limits.maxSampledTexturesPerShaderStage`.

**Comment by Zyie** [comment]
Hey @dev7355608 just for future reference fixes go into `main` and features into `dev`!

**Comment by dev7355608** [comment]
Ah, that makes sense. I thought I targeted the wrong branch with all my PRs.

**Pull request merged**